src/trace_ng/log.zig

```diff
@@ -0,0 +1,273 @@
+const testing = std.testing;
+const Allocator = std.mem.Allocator;
+const AtomicBool = std.atomic.Value(bool);
+const RecycleFBA = sig.utils.allocators.RecycleFBA;
+
+pub const LogConfig = struct {
+    level: Level = Level.debug,
+    buff_size: usize = 64,
+};
+
+pub const Config = struct {
+    level: Level = Level.debug,
+    allocator: std.mem.Allocator,
+    fba_bytes: u64,
+};
+
+const INITIAL_LOG_CHANNEL_SIZE: usize = 1024;
+
+const UnscopedLogger = StandardLogger(null);
+pub fn StandardLogger(comptime scope: ?type) type {
+    return struct {
+        const Self = @This();
+        level: Level,
+        exit_sig: AtomicBool,
+        allocator: Allocator,
+        free_fba: RecycleFBA,
+        fba_bytes: u64,
+        channel: *Channel(logfmt.LogMsg),
+        handle: ?std.Thread,
+
+        pub fn init(config: Config) Self {
+            return .{
+                .allocator = config.allocator,
+                .free_fba = RecycleFBA.init(config.allocator, config.fba_bytes) catch @panic("could not create RecycleFBA"),
+                .fba_bytes = config.fba_bytes,
+                .level = config.level,
+                .exit_sig = AtomicBool.init(false),
+                .channel = Channel(logfmt.LogMsg).init(config.allocator, INITIAL_LOG_CHANNEL_SIZE),
+                .handle = null,
+            };
+        }
+
+        fn unscoped(self: Self) UnscopedLogger {
+            return .{
+                .level = self.level,
+                .exit_sig = self.exit_sig,
+                .allocator = self.allocator,
+                .free_fba = self.free_fba,
+                .fba_bytes = self.fba_bytes,
+                .channel = self.channel,
+                .handle = self.handle,
+            };
+        }
+
+        fn withScope(self: Self, comptime new_scope: anytype) StandardLogger(new_scope) {
+            return .{
+                .level = self.level,
+                .exit_sig = self.exit_sig,
+                .allocator = self.allocator,
+                .free_fba = self.free_fba,
+                .fba_bytes = self.fba_bytes,
+                .channel = self.channel,
+                .handle = self.handle,
+            };
+        }
+
+        pub fn spawn(self: *Self) void {
+            self.handle = std.Thread.spawn(.{}, Self.run, .{self}) catch @panic("could not spawn Logger");
+        }
+
+        pub fn deinit(self: *Self) void {
+            self.channel.close();
+            if (self.handle) |handle| {
+                self.exit_sig.store(true, .seq_cst);
+                handle.join();
+            }
+            self.channel.deinit();
+            self.free_fba.deinit();
+        }
+
+        pub fn run(self: *Self) void {
+            while (!self.exit_sig.load(.seq_cst)) {
+                std.time.sleep(std.time.ns_per_ms * 5);
+                const messages = self.channel.drain() orelse {
+                    // channel is closed
+                    return;
```

Comment:
You can use `std.fmt.count` to determine the number of bytes you need to allocate.

Reply:
Used here https://github.com/Syndica/sig/pull/249/files#diff-a5f543f6d9dfe65e2e547cdf7bd2cd49d8e85d4afccc81a3c51568b0208dfeb7R29